# Genomic Signatures

```{r signatures_load}
variant_caller <- "HaplotypeCaller" # only using this caller for signatures right now, should be embedded in filename
# filenames: sampleID.HaplotypeCaller.signatures.Rds
signatures_Rds <- paste(sampleID, variant_caller, params$signatures_Rds, sep = '.')[1] 
signatures_plot_Rds <- paste(sampleID, variant_caller, params$signatures_plot_Rds, sep = '.')[1]
signatures_pieplot_Rds <- paste(sampleID, variant_caller, params$signatures_pieplot_Rds, sep = '.')[1]

# start with an emtpy list, try to load the signatures into it
signatures <- list()
tryCatch(
        {
            signatures[["signatures"]] <- readRDS(signatures_Rds)
            signatures[["signatures_plot"]] <- readRDS(signatures_plot_Rds)
            signatures[["signatures_pieplot"]] <- readRDS(signatures_pieplot_Rds)
            
            weights <- as.data.frame(t(signatures[["signatures"]][["weights"]]))
            weights <- weights[ order(-weights[,1]), , drop=FALSE]
            
            signatures[["weights"]] <- weights
        }
    )

```

- variant caller: `r variant_caller`

```{r signatures_weights, results='asis'}
if (length(signatures) < 1){
    cat('\n\n- __No results found__\n\n')
} else {
    cat(sprintf('\n\n## %s\n\n', "Weights"))
    signatures[["weights"]]
}

```

## Plots

```{r, results='asis', warning = FALSE}
if (length(signatures) < 1){
    cat('\n\n- __No results found__\n\n')
} else {
    print(signatures[["signatures_pieplot"]])
}

```


```{r, results='asis', warning = FALSE}
if (length(signatures) < 1){
    cat('\n\n- __No results found__\n\n')
} else {
    print(signatures[["signatures_plot"]])
}

```


```{r}
save.image("signatures.Rdata", compress = TRUE)
```