```{r}
# remove_vars(vars = c(
#     "haplotypecaller_files",
#     "haplotypecaller_df",
#     "make_haplotypecaller_plots",
#     "haplotypecaller_plots"
#     ))

# ~~~~~~~ FUNCTIONS ~~~~~~~ #
make_haplotypecaller_plots <- function(df){
    plots <- list()
    if(is.data.frame(df)){
        if(nrow(df) > 0){
            # ~~~~~~~~~~~~~~~~~~~~~~~~~ # 
            haplotypecaller_counts <- ggplot(data = df, aes(x = Sample)) +
                geom_bar(stat="count") + 
                ggtitle("Number of Mutations") + 
                coord_flip() +
                theme_bw() +
                theme(panel.grid.minor = element_blank()) 
            
            plots[['counts']] <- list(
                plot = haplotypecaller_counts,
                subheader = 'Counts'
            )
            # ~~~~~~~~~~~~~~~~~~~~~~~~~ # 
            haplotypecaller_qual <- ggplot(data = df, aes(x=Sample, y=QUAL, fill=Sample)) + 
                geom_boxplot() + 
                guides(fill=FALSE) + 
                coord_flip() + 
                ggtitle("Quality Scores") + 
                theme_bw() +
                theme(panel.grid.minor = element_blank())
            
            plots[['qual']] <- list(
                plot = haplotypecaller_qual,
                subheader = 'Quality Scores'
            )
            # ~~~~~~~~~~~~~~~~~~~~~~~~~ # 
            haplotypecaller_dp <- ggplot(data = df, aes(x=Sample, y=DP, fill=Sample)) + 
                geom_boxplot() + 
                guides(fill=FALSE) + 
                coord_flip() + 
                ggtitle("Depth of Coverage") + 
                theme_bw() +
                theme(panel.grid.minor = element_blank())
            
            plots[['dp']] <- list(
                plot = haplotypecaller_dp,
                subheader = 'Quality Scores'
            )
            # ~~~~~~~~~~~~~~~~~~~~~~~~~ # 
            haplotypecaller_af <- ggplot(data = df, aes(x=Sample, y=AF, fill=Sample)) + 
                geom_boxplot() + 
                guides(fill=FALSE) + 
                coord_flip() + 
                ggtitle("Allele Frequency") + 
                theme_bw() +
                theme(panel.grid.minor = element_blank())
            
            plots[['af']] <- list(
                plot = haplotypecaller_af,
                subheader = 'Allele Frequency'
            )
        }
    }
    return(plots)
}


# find matching files; missing or no-match files dropped by default
haplotypecaller_files <- unlist(sapply(X = samplesIDs, FUN = function(x){
    pattern <- paste(x, '.HaplotypeCaller.annotations.tsv', sep = '')
    return(input_files[grep(pattern = pattern, x = basename(input_files))])
}))

haplotypecaller_df <- do.call('rbind', lapply(X = haplotypecaller_files, FUN = function(x){
    df <- read.delim(file = x, header = TRUE, sep = '\t')
    return(df)
}))

haplotypecaller_plots <- make_haplotypecaller_plots(haplotypecaller_df)
```


# GATK HaplotypeCaller

Variant calling results

```{r, results='asis'}
# print any results found
if (length(haplotypecaller_plots) < 1){
    cat('\n\n- __No results found__\n\n')
} else {
    for(item in haplotypecaller_plots){
        subplot<- item[["plot"]]
        subheader <- item[["subheader"]]
        
        cat(sprintf('\n\n## %s\n\n', subheader))
        print(subplot)
    }
}
```

