```{r load_reads}
# minimum number of reads required (millions)
min_reads_cutoff <- 10

# minimum number of reads for NTC control sample
min_reads_ntc <- 30000

# BWA original flagstat values
flagstat_df <- read.delim(file = params$flagstat_table, header = TRUE, sep = '\t') 
flagstat_df[["MappedReads"]] <- flagstat_df[["SequencedPairedReads"]]

# sambamaba/samtools dedup flagstat values
dedup_flagstat_df <- read.delim(file = params$dedup_flagstat_table, header = TRUE, sep = '\t')
dedup_flagstat_df[["DeduplicatedReads"]] <- dedup_flagstat_df[["SequencedPairedReads"]]

# sambamaba/samtools dedup log message values
dedup_df <- read.delim(file = params$reads_dedup_table, header = TRUE, sep = '\t')
dedup_df[["DuplicatedReads"]] <- dedup_df[["Duplicates"]]

# create aggregate df
reads_df <- merge(x = flagstat_df[, c("MappedReads", "Sample")], y = dedup_df[, c("DuplicatedReads", "Sample")], by = "Sample")
reads_df <- merge(x = reads_df, y = dedup_flagstat_df[, c("DeduplicatedReads", "Sample")], by = "Sample")

# sort Sample factor levels
reads_df[["Sample"]] <- factor(x = reads_df[["Sample"]], levels = sort(unique(as.character(reads_df[["Sample"]]), decreasing = TRUE)))

# convert to long format for plotting
reads_df_long <- reshape2::melt(reads_df, id.vars="Sample", variable.name="Type", value.name="Reads")

# convert to millions for plot
reads_df_long[["Reads"]] <- as.numeric(reads_df_long[["Reads"]]) / 1e6

# sort Type factor levels
reads_df_long[["Type"]] <- factor(x = reads_df_long[["Type"]], levels = sort(unique(as.character(reads_df_long[["Type"]]), decreasing = TRUE)))

mapping_plot <- ggplot(data = reads_df_long[ reads_df_long[["Type"]] %in% c("MappedReads", "DeduplicatedReads") , ], 
       aes(x = Sample, y = Reads, fill = Type)) + 
    geom_bar(stat="identity", position = "dodge") +
    geom_hline(yintercept = min_reads_cutoff, color = "red") +
    ggtitle("Sample Read Mapping") +
    coord_flip() +
    ylab("Reads (millions)") +
    theme_bw() +
    theme(panel.grid.minor = element_blank())
    
```

# Sequencing Reads

Genomic sequencing read alignment and deduplication results. Reads are aligned with BWA [@BWAMEM] and deduplicated with Sambamba [@Sambamba]. Unique deduplicated reads are used for downstream analysis.

```{r print_reads, results='asis'}
DT::datatable(data = reads_df,
              options = list(pageLength = nrow(reads_df)),
              class = 'cell-border stripe')
```

- No Template Control sample (NTC) must have fewer than `r as.character(min_reads_ntc)` reads for sequencing run to qualify for NGS580 analysis. Large amounts of reads in this sample suggest potential contamination. 

```{r mapping_plot}
mapping_plot
```

- Each sample must have at least `r min_reads_cutoff` million deduplicated reads to qualify for NGS580 panel analysis (red line on plot).

```{r}
save.image(file="reads.Rdata", compress = TRUE)
```
