# QC: SeraCare Pool Comparision

SeraCare is a positive controle synthetic variant DNA sample that may be included in sequencing runs to ensure that variants are being detected properly. The variants detected in the current run's SeraCare sample may be compared to a background pool of previously sequenced SeraCare samples to determine if variants are being called with the expected frequency.

```{r}
seracare_annotations_file <- params$seracare_annotations_file # current run SeraCare sample variants
seracare_pool_annotations_file <- params$seracare_pool_annotations_file # SeraCare pool variants
seracare_selected_variants <- params$seracare_selected_variants # list of known SeraCare variants

seracare_annotations <- data.frame()
seracare_pool_annotations <- data.frame()
seracare_selected <- data.frame()

# file may not have been produced if no SeraCare samples were included in run; 
# try to load file if exists
tryCatch(
    {
        seracare_annotations <- read.delim(file = seracare_annotations_file, 
                                  header = TRUE, 
                                  sep = '\t', 
                                  check.names = FALSE, 
                                  na.strings = c('.'))
        seracare_pool_annotations <- read.delim(file = seracare_pool_annotations_file, 
                                  header = TRUE, 
                                  sep = '\t', 
                                  check.names = FALSE, 
                                  na.strings = c('.'))
        seracare_selected <- read.delim(file = seracare_selected_variants, 
                                  header = TRUE, 
                                  sep = '\t', 
                                  check.names = FALSE, 
                                  na.strings = c('.'))
        
        # sort the Sample levels
        seracare_annotations[["Sample"]] <- factor(x = seracare_annotations[["Sample"]],
                                         levels = sort(unique(as.character(seracare_annotations[["Sample"]]),
                                                              decreasing = TRUE)))
        seracare_annotations[["Mut_ID"]] <- sprintf("%s %s", seracare_annotations[["SeraCare.Gene"]], seracare_annotations[["SeraCare.AminoAcid"]])
        # convert expect AF to factor level
        seracare_annotations[["SeraCare.TargetAF"]] <- factor(x = seracare_annotations[["SeraCare.TargetAF"]], levels = sort(unique(seracare_annotations[["SeraCare.TargetAF"]])))
    },
    error=function(cond) {
        return(NA)
    }
)

```

## LoFreq

SeraCare filtered variants called by LoFreq.

```{r}
if(nrow(seracare_annotations) > 0){
    
    print_cols <- c("Gene.refGene", "SeraCare.AminoAcid", "DP", "AF","SeraCare.TargetAF", 
                    "Chr", "Start", "End", "Ref", "Alt", 
                     "Sample", "SeraCare.COSMIC")
    DT::datatable(data = seracare_annotations[ , print_cols],
                  options = list(pageLength = 25),
                  class = 'cell-border stripe')
    
    ggplot(data =  seracare_annotations, 
           aes(y = AF, 
               x = Sample, 
               color = SeraCare.TargetAF, 
               group = Mut_ID) ) + 
        geom_point() +
        geom_line() +
        theme_bw() +
        theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
        # scale_y_continuous(limits = c(0, NA), 
        #                    breaks = seq(0, max(seracare_annotations[["AF"]]), 0.025)) +
        # facet_grid(~SeraCare.TargetAF) +
        ylab("Detected Variant Allele Frequency") +
        ggtitle(sprintf('SeraCare Variant Frequencies'))
    
} else {
    cat('\n\n- No results found')
}
```

```{r}
save.image(file="seracare.pool.Rdata",compress = TRUE)
```