```{r load_annotations}
source("util/karyotype.R")
# get base karyotype plot objects
karyotype <- karyotype_plot()
chrom_order <- get_chrom_order()

# loads variant information from 'all_annotations.tsv' file
annotations_file <- params$annotations_file
annotations <- read.delim(file = annotations_file, 
                          header = TRUE, 
                          sep = '\t', 
                          check.names = FALSE, 
                          na.strings = c('.'))

save.image(file="variants.loaded.Rdata",compress = TRUE)

# sort the Sample levels
annotations[["Sample"]] <- factor(x = annotations[["Sample"]], 
                                  levels = sort(unique(as.character(annotations[["Sample"]]), 
                                                       decreasing = TRUE)))

# add a sorted chromosome column for karyotype
annotations[["chromosome"]] <- factor(x = annotations[["Chr"]], 
                                          levels = chrom_order)

# add '1' column for aggregations
annotations[["n"]] <- 1
```

# Variants

## Summary Metrics

### Variants per Sample

```{r}
DT::datatable(data = aggregate(n ~ Sample, data =annotations, FUN = sum),
              options = list(pageLength = nrow(aggregate(n ~ Sample, data =annotations, FUN = sum))),
              class = 'cell-border stripe')

```

### Variants per Variant Caller

```{r}
DT::datatable(data = aggregate(n ~ VariantCaller, data =annotations, FUN = sum),
              options = list(pageLength = nrow(aggregate(n ~ VariantCaller, data =annotations, FUN = sum))),
              class = 'cell-border stripe')

```

### Variants per Type

```{r}
DT::datatable(data = aggregate(n ~ Func.refGene, data =annotations, FUN = sum),
              options = list(pageLength = nrow(aggregate(n ~ Func.refGene, data =annotations, FUN = sum))),
              class = 'cell-border stripe')

```

### Variants per Gene

```{r}
DT::datatable(data = aggregate(n ~ Gene.refGene, data =annotations, FUN = sum),
              options = list(pageLength = 10),
              class = 'cell-border stripe')
```


## MuTect2

Somatic variant calling with GATK MuTect2 [@GATKToolkit], using matched tumor-normal pairs.

```{r mutect2}
make_mutect2_plots <- function(df){
    plots <- list()
    if(is.data.frame(df)){
        if(nrow(df) > 0){
            # ~~~~~~~~~~ VARIANT TYPES PIE CHART ~~~~~~~~~~~~~~~ # 
            variant_types_pie <- ggplot(aggregate( n ~ Func.refGene, data = df, FUN = sum), 
                                       aes(x= "", y = n, fill = Func.refGene)) + 
                geom_bar(stat = "identity", width = 1) + # , color = "white"
                coord_polar("y", start = 0) +
                theme_bw() +
                theme(panel.grid.minor = element_blank(), 
                      panel.grid.major = element_blank(), 
                      panel.border = element_blank(), 
                      axis.ticks = element_blank(), 
                      axis.text = element_blank(), 
                      axis.title = element_blank()) +
                ggtitle("Target Types")
            
            plots[['types']] <- list(
                plot = variant_types_pie,
                subheader = 'Variant Types'
            )
            # ~~~~~~~~~~~ VARIANT COUNT BARPLOT ~~~~~~~~~~~~~~ # 
            mutect2_counts <- ggplot(data = df, aes(x = Sample)) +
                geom_bar(stat="count") + 
                ggtitle("Number of Variants") + 
                coord_flip() +
                theme_bw() +
                theme(panel.grid.minor = element_blank()) 
            
            plots[['counts']] <- list(
                plot = mutect2_counts,
                subheader = 'Counts'
            )
            # ~~~~~~~~~~~ QUALITY SCORE BOXPLOT ~~~~~~~~~~~~~~ # 
            mutect2_qual <- ggplot(data = df, aes(x=Sample, y=QUAL, fill=Sample)) + 
                geom_boxplot() + 
                # cut off x axis at 1000
                coord_cartesian( ylim = c(0, 1000)) +
                guides(fill=FALSE) + 
                coord_flip() + 
                ggtitle("Quality Scores") + 
                theme_bw() +
                theme(panel.grid.minor = element_blank())
            
            plots[['qual']] <- list(
                plot = mutect2_qual,
                subheader = 'Quality Scores'
            )
            # ~~~~~~~~~~~ VARIANT DEPTH BOXPLOT ~~~~~~~~~~~~~~ # 
            mutect2_dp <- ggplot(data = df, aes(x=Sample, y=TUMOR.AD.TOTAL, fill=Sample)) + 
                geom_boxplot() + 
                guides(fill=FALSE) + 
                coord_flip() + 
                ggtitle("Tumor Depth of Coverage") + 
                theme_bw() +
                theme(panel.grid.minor = element_blank())
            
            plots[['dp']] <- list(
                plot = mutect2_dp,
                subheader = 'Tumor Depth of Coverage'
            )
            # ~~~~~~~~~~ VARIANT FREQUENCY BOXPLOT ~~~~~~~~~~~~~~~ # 
            mutect2_af <- ggplot(data = df, aes(x=Sample, y=TUMOR.AF, fill=Sample)) + 
                geom_boxplot() + 
                guides(fill=FALSE) + 
                coord_flip() + 
                ggtitle("Tumor Allele Frequency") + 
                theme_bw() +
                theme(panel.grid.minor = element_blank())
            
            plots[['af']] <- list(
                plot = mutect2_af,
                subheader = 'Tumor Allele Frequency'
            )
            # ~~~~~~~~~~ VARIANT KARYOTYPE ~~~~~~~~~~~~~~~ # 
            mutect2_kar <- karyotype + 
                geom_point(data = df, 
                           aes(x = as.numeric(chromosome),
                               y = Start), color = "red") +
                ggtitle("MuTect2 Variant Karotype") +
                theme(legend.position="none")
            
            plots[['kar']] <- list(
                plot = mutect2_kar,
                subheader = 'Variant Karotype'
            )
        }
    }
    return(plots)
}

mutect2_plots <- make_mutect2_plots(annotations[annotations[["VariantCaller"]] == 'MuTect2', ])
```

```{r mutect2_print, results='asis'}
# print any results found
if (length(mutect2_plots) < 1){
    cat('\n\n- __No results found__\n\n')
} else {
    for(item in mutect2_plots){
        subplot<- item[["plot"]]
        subheader <- item[["subheader"]]
        
        cat(sprintf('\n\n### %s\n\n', subheader))
        print(subplot)
    }
}
```

## LoFreq Somatic

Somatic variant calling with LoFreq [@LoFreq], using matched tumor-normal pairs. 


```{r}
# ~~~~~~~ FUNCTIONS ~~~~~~~ #
make_lofreq_somatic_plots <- function(df){
    plots <- list()
    if(is.data.frame(df)){
        if(nrow(df) > 0){
            # ~~~~~~~~~~ VARIANT TYPES PIE CHART ~~~~~~~~~~~~~~~ # 
            variant_types_pie <- ggplot(aggregate( n ~ Func.refGene, data = df, FUN = sum), 
                                       aes(x= "", y = n, fill = Func.refGene)) + 
                geom_bar(stat = "identity", width = 1) + # , color = "white"
                coord_polar("y", start = 0) +
                theme_bw() +
                theme(panel.grid.minor = element_blank(), 
                      panel.grid.major = element_blank(), 
                      panel.border = element_blank(), 
                      axis.ticks = element_blank(), 
                      axis.text = element_blank(), 
                      axis.title = element_blank()) +
                ggtitle("Target Types")
            
            plots[['types']] <- list(
                plot = variant_types_pie,
                subheader = 'Variant Types'
            )
            # ~~~~~~~~~~~ VARIANT COUNT BARPLOT ~~~~~~~~~~~~~~ # 
            lofreq_counts <- ggplot(data = df, aes(x = Sample)) +
                geom_bar(stat="count") + 
                ggtitle("Number of Variants") + 
                coord_flip() +
                theme_bw() +
                theme(panel.grid.minor = element_blank()) 
            
            plots[['counts']] <- list(
                plot = lofreq_counts,
                subheader = 'Counts'
            )
            # ~~~~~~~~~~~~~~~~~~~~~~~~~ # 
            lofreq_qual <- ggplot(data = df, aes(x=Sample, y=QUAL, fill=Sample)) + 
                geom_boxplot() + 
                guides(fill=FALSE) + 
                coord_flip() + 
                ggtitle("Quality Scores") + 
                theme_bw() +
                theme(panel.grid.minor = element_blank())
            
            plots[['qual']] <- list(
                plot = lofreq_qual,
                subheader = 'Quality Scores'
            )
            # ~~~~~~~~~~~~~~~~~~~~~~~~~ # 
            lofreq_dp <- ggplot(data = df, aes(x=Sample, y=DP, fill=Sample)) + 
                geom_boxplot() + 
                guides(fill=FALSE) + 
                coord_flip() + 
                ggtitle("Depth of Coverage") + 
                theme_bw() +
                theme(panel.grid.minor = element_blank())
            
            plots[['dp']] <- list(
                plot = lofreq_dp,
                subheader = 'Quality Scores'
            )
            # ~~~~~~~~~~~~~~~~~~~~~~~~~ # 
            lofreq_af <- ggplot(data = df, aes(x=Sample, y=AF, fill=Sample)) + 
                geom_boxplot() + 
                guides(fill=FALSE) + 
                coord_flip() + 
                ggtitle("Allele Frequency") + 
                theme_bw() +
                theme(panel.grid.minor = element_blank())
            
            plots[['af']] <- list(
                plot = lofreq_af,
                subheader = 'Allele Frequency'
            )
            # ~~~~~~~~~~ VARIANT KARYOTYPE ~~~~~~~~~~~~~~~ # 
            lofreq_kar <- karyotype + 
                geom_point(data = df, 
                           aes(x = as.numeric(chromosome),
                               y = Start), color = "red") +
                ggtitle("Lofreq Somatic Variant Karotype") +
                theme(legend.position="none")
            
            plots[['kar']] <- list(
                plot = lofreq_kar,
                subheader = 'Variant Karotype'
            )
        }
    }
    return(plots)
}

# ~~~~~~~ RUN ~~~~~~~ #
lofreq_somatic_plots <- make_lofreq_somatic_plots(annotations[annotations[["VariantCaller"]] == 'LoFreqSomatic', ])
```

```{r, results='asis'}
# print any results found
if (length(lofreq_somatic_plots) < 1){
    cat('\n\n- __No results found__\n\n')
} else {
    
    for(item in lofreq_somatic_plots){
        subplot<- item[["plot"]]
        subheader <- item[["subheader"]]
        
        cat(sprintf('\n\n### %s\n\n', subheader))
        print(subplot)
    }
}
```


## GATK HaplotypeCaller

Variant calling with GATK HaplotypeCaller [@GATKToolkit].

```{r haplotypecaller}
# ~~~~~~~ FUNCTIONS ~~~~~~~ #
make_haplotypecaller_plots <- function(df){
    plots <- list()
    if(is.data.frame(df)){
        if(nrow(df) > 0){
            # ~~~~~~~~~~ VARIANT TYPES PIE CHART ~~~~~~~~~~~~~~~ # 
            variant_types_pie <- ggplot(aggregate( n ~ Func.refGene, data = df, FUN = sum), 
                                       aes(x= "", y = n, fill = Func.refGene)) + 
                geom_bar(stat = "identity", width = 1) + # , color = "white"
                coord_polar("y", start = 0) +
                theme_bw() +
                theme(panel.grid.minor = element_blank(), 
                      panel.grid.major = element_blank(), 
                      panel.border = element_blank(), 
                      axis.ticks = element_blank(), 
                      axis.text = element_blank(), 
                      axis.title = element_blank()) +
                ggtitle("Target Types")
            
            plots[['types']] <- list(
                plot = variant_types_pie,
                subheader = 'Variant Types'
            )
            # ~~~~~~~~~~~~ VARIANT COUNT BARPLOT ~~~~~~~~~~~~~ # 
            haplotypecaller_counts <- ggplot(data = df, aes(x = Sample)) +
                geom_bar(stat="count") + 
                ggtitle("Number of Variants") + 
                coord_flip() +
                theme_bw() +
                theme(panel.grid.minor = element_blank()) 
            
            plots[['counts']] <- list(
                plot = haplotypecaller_counts,
                subheader = 'Counts'
            )
            # ~~~~~~~~~~~~~~~~~~~~~~~~~ # 
            haplotypecaller_qual <- ggplot(data = df, aes(x=Sample, y=QUAL, fill=Sample)) + 
                geom_boxplot() + 
                guides(fill=FALSE) + 
                coord_flip() + 
                ggtitle("Quality Scores") + 
                theme_bw() +
                theme(panel.grid.minor = element_blank())
            
            plots[['qual']] <- list(
                plot = haplotypecaller_qual,
                subheader = 'Quality Scores'
            )
            # ~~~~~~~~~~~~~~~~~~~~~~~~~ # 
            haplotypecaller_dp <- ggplot(data = df, aes(x=Sample, y=DP, fill=Sample)) + 
                geom_boxplot() + 
                guides(fill=FALSE) + 
                coord_flip() + 
                ggtitle("Depth of Coverage") + 
                theme_bw() +
                theme(panel.grid.minor = element_blank())
            
            plots[['dp']] <- list(
                plot = haplotypecaller_dp,
                subheader = 'Depth of Coverage'
            )
            # ~~~~~~~~~~~~~~~~~~~~~~~~~ # 
            haplotypecaller_af <- ggplot(data = df, aes(x=Sample, y=AF, fill=Sample)) + 
                geom_boxplot() + 
                guides(fill=FALSE) + 
                coord_flip() + 
                ggtitle("Allele Frequency") + 
                theme_bw() +
                theme(panel.grid.minor = element_blank())
            
            plots[['af']] <- list(
                plot = haplotypecaller_af,
                subheader = 'Allele Frequency'
            )
            # ~~~~~~~~~~ VARIANT KARYOTYPE ~~~~~~~~~~~~~~~ # 
            haplotypecaller_kar <- karyotype + 
                geom_point(data = df, 
                           aes(x = as.numeric(chromosome),
                               y = Start), color = "red") +
                ggtitle("GATK HaplotypeCaller Variant Karotype") +
                theme(legend.position="none")
            
            plots[['kar']] <- list(
                plot = haplotypecaller_kar,
                subheader = 'Variant Karotype'
            )
        }
    }
    return(plots)
}

# ~~~~~~~ RUN ~~~~~~~ #
haplotypecaller_plots <- make_haplotypecaller_plots(annotations[annotations[["VariantCaller"]] == 'HaplotypeCaller', ])
```

```{r haplotypecaller_print, results='asis'}
# print any results found
if (length(haplotypecaller_plots) < 1){
    cat('\n\n- __No results found__\n\n')
} else {
    for(item in haplotypecaller_plots){
        subplot<- item[["plot"]]
        subheader <- item[["subheader"]]
        
        cat(sprintf('\n\n### %s\n\n', subheader))
        print(subplot)
    }
}
```

## LoFreq

Variant calling results from the LoFreq program [@LoFreq].

```{r lofreq}
# ~~~~~~~ FUNCTIONS ~~~~~~~ #
make_lofreq_plots <- function(df){
    plots <- list()
    if(is.data.frame(df)){
        if(nrow(df) > 0){
            # ~~~~~~~~~~ VARIANT TYPES PIE CHART ~~~~~~~~~~~~~~~ # 
            variant_types_pie <- ggplot(aggregate( n ~ Func.refGene, data = df, FUN = sum), 
                                       aes(x= "", y = n, fill = Func.refGene)) + 
                geom_bar(stat = "identity", width = 1) + # , color = "white"
                coord_polar("y", start = 0) +
                theme_bw() +
                theme(panel.grid.minor = element_blank(), 
                      panel.grid.major = element_blank(), 
                      panel.border = element_blank(), 
                      axis.ticks = element_blank(), 
                      axis.text = element_blank(), 
                      axis.title = element_blank()) +
                ggtitle("Target Types")
            
            plots[['types']] <- list(
                plot = variant_types_pie,
                subheader = 'Variant Types'
            )
            # ~~~~~~~~~~ VARIANT COUNT BARPLOT ~~~~~~~~~~~~~~~ # 
            lofreq_counts <- ggplot(data = df, aes(x = Sample)) +
                geom_bar(stat="count") + 
                ggtitle("Number of Variants") + 
                coord_flip() +
                theme_bw() +
                theme(panel.grid.minor = element_blank()) 
            
            plots[['counts']] <- list(
                plot = lofreq_counts,
                subheader = 'Counts'
            )
            # ~~~~~~~~~~~~~~~~~~~~~~~~~ # 
            lofreq_qual <- ggplot(data = df, aes(x=Sample, y=QUAL, fill=Sample)) + 
                geom_boxplot() + 
                guides(fill=FALSE) + 
                coord_flip() + 
                ggtitle("Quality Scores") + 
                theme_bw() +
                theme(panel.grid.minor = element_blank())
            
            plots[['qual']] <- list(
                plot = lofreq_qual,
                subheader = 'Quality Scores'
            )
            # ~~~~~~~~~~~~~~~~~~~~~~~~~ # 
            lofreq_dp <- ggplot(data = df, aes(x=Sample, y=DP, fill=Sample)) + 
                geom_boxplot() + 
                guides(fill=FALSE) + 
                coord_flip() + 
                ggtitle("Depth of Coverage") + 
                theme_bw() +
                theme(panel.grid.minor = element_blank())
            
            plots[['dp']] <- list(
                plot = lofreq_dp,
                subheader = 'Depth of Coverage'
            )
            # ~~~~~~~~~~~~~~~~~~~~~~~~~ # 
            lofreq_af <- ggplot(data = df, aes(x=Sample, y=AF, fill=Sample)) + 
                geom_boxplot() + 
                guides(fill=FALSE) + 
                coord_flip() + 
                ggtitle("Allele Frequency") + 
                theme_bw() +
                theme(panel.grid.minor = element_blank())
            
            plots[['af']] <- list(
                plot = lofreq_af,
                subheader = 'Allele Frequency'
            )
            # ~~~~~~~~~~ VARIANT KARYOTYPE ~~~~~~~~~~~~~~~ # 
            lofreq_kar <- karyotype + 
                geom_point(data = df, 
                           aes(x = as.numeric(chromosome),
                               y = Start), color = "red") +
                ggtitle("Lofreq Variant Karotype") +
                theme(legend.position="none")
            
            plots[['kar']] <- list(
                plot = lofreq_kar,
                subheader = 'Variant Karotype'
            )
        }
    }
    return(plots)
}

# ~~~~~~~ RUN ~~~~~~~ #
lofreq_plots <- make_lofreq_plots(annotations[annotations[["VariantCaller"]] == 'LoFreq', ])
```

```{r plot_variants, results='asis'}
# print any results found
if (length(lofreq_plots) < 1){
    cat('\n\n- __No results found__\n\n')
} else {
    
    for(item in lofreq_plots){
        subplot<- item[["plot"]]
        subheader <- item[["subheader"]]
        
        cat(sprintf('\n\n### %s\n\n', subheader))
        print(subplot)
    }
}
```

## VarScan2 SNP

Variant calling with VarScan2 [@VarScan2] (SNP's).

```{r varscan_snp}
# ~~~~~~~ FUNCTIONS ~~~~~~~ #
make_varscan2_plots <- function(df){
    plots <- list()
    if(is.data.frame(df)){
        if(nrow(df) > 0){
            
            # ~~~~~~~~~~ VARIANT TYPES PIE CHART ~~~~~~~~~~~~~~~ # 
            variant_types_pie <- ggplot(aggregate( n ~ Func.refGene, data = df, FUN = sum), 
                                       aes(x= "", y = n, fill = Func.refGene)) + 
                geom_bar(stat = "identity", width = 1) + # , color = "white"
                coord_polar("y", start = 0) +
                theme_bw() +
                theme(panel.grid.minor = element_blank(), 
                      panel.grid.major = element_blank(), 
                      panel.border = element_blank(), 
                      axis.ticks = element_blank(), 
                      axis.text = element_blank(), 
                      axis.title = element_blank()) +
                ggtitle("Target Types")
            
            plots[['types']] <- list(
                plot = variant_types_pie,
                subheader = 'Variant Types'
            )
            # ~~~~~~~~~~ VARIANT COUNT BARPLOT ~~~~~~~~~~~~~~~ # 
            counts_barplot <- ggplot(data = df, aes(x = Sample)) +
                geom_bar(stat="count") + 
                ggtitle("Number of Variants") + 
                coord_flip() +
                theme_bw() +
                theme(panel.grid.minor = element_blank()) 
            
            plots[['counts']] <- list(
                plot = counts_barplot,
                subheader = 'Counts'
            )
            # ~~~~~~~~~~~~~~~~~~~~~~~~~ # 
            qual_boxplot1 <- ggplot(data = df, aes(x=Sample, y=ABQ, fill=Sample)) + 
                geom_boxplot() + 
                guides(fill=FALSE) + 
                coord_flip() + 
                ggtitle("Variant Base Quality Scores") + 
                theme_bw() +
                theme(panel.grid.minor = element_blank())
            
            plots[['qual1']] <- list(
                plot = qual_boxplot1,
                subheader = 'Average quality of variant-supporting bases'
            )
            # ~~~~~~~~~~~~~~~~~~~~~~~~~ # 
            qual_boxplot2 <- ggplot(data = df, aes(x=Sample, y=RBQ, fill=Sample)) + 
                geom_boxplot() + 
                guides(fill=FALSE) + 
                coord_flip() + 
                ggtitle("Reference Base Quality Scores") + 
                theme_bw() +
                theme(panel.grid.minor = element_blank())
            
            plots[['qual2']] <- list(
                plot = qual_boxplot2,
                subheader = 'Average quality of reference-supporting bases'
            )
            # ~~~~~~~~~~~~~~~~~~~~~~~~~ # 
            depth_boxplot <- ggplot(data = df, aes(x=Sample, y=DP, fill=Sample)) + 
                geom_boxplot() + 
                guides(fill=FALSE) + 
                coord_flip() + 
                ggtitle("Depth of Coverage") + 
                theme_bw() +
                theme(panel.grid.minor = element_blank())
            
            plots[['dp']] <- list(
                plot = depth_boxplot,
                subheader = 'Depth of Coverage'
            )
            # ~~~~~~~~~~~~~~~~~~~~~~~~~ # 
            frequency_boxplot <- ggplot(data = df, aes(x=Sample, y=FREQ, fill=Sample)) + 
                geom_boxplot() + 
                guides(fill=FALSE) + 
                coord_flip() + 
                ggtitle("Allele Frequency") + 
                theme_bw() +
                theme(panel.grid.minor = element_blank())
            
            plots[['af']] <- list(
                plot = frequency_boxplot,
                subheader = 'Allele Frequency'
            )
            # ~~~~~~~~~~ VARIANT KARYOTYPE ~~~~~~~~~~~~~~~ # 
            kar <- karyotype + 
                geom_point(data = df, 
                           aes(x = as.numeric(chromosome),
                               y = Start), color = "red") +
                ggtitle("VarScan2 Variant Karotype") +
                theme(legend.position="none")
            
            plots[['kar']] <- list(
                plot = kar,
                subheader = 'Variant Karotype'
            )
        }
    }
    return(plots)
}

# ~~~~~~~ RUN ~~~~~~~ #
varscan2_snp_plots <- make_varscan2_plots(annotations[which(annotations[["VariantCaller"]] == 'VarScan2' & annotations[["VariantCallerType"]] == "snp" ), ])


```

```{r, results='asis'}
# print any results found
if (length(varscan2_snp_plots) < 1){
    cat('\n\n- __No results found__\n\n')
} else {
    
    for(item in varscan2_snp_plots){
        subplot<- item[["plot"]]
        subheader <- item[["subheader"]]
        
        cat(sprintf('\n\n### %s\n\n', subheader))
        print(subplot)
    }
}
```

## VarScan2 Indel

Variant calling with VarScan2 [@VarScan2] (indels).

```{r varscan_indel}
# ~~~~~~~ FUNCTIONS ~~~~~~~ #
make_varscan2_indel_plots <- function(df){
    plots <- list()
    if(is.data.frame(df)){
        if(nrow(df) > 0){
            
            # ~~~~~~~~~~ VARIANT TYPES PIE CHART ~~~~~~~~~~~~~~~ # 
            variant_types_pie <- ggplot(aggregate( n ~ Func.refGene, data = df, FUN = sum), 
                                       aes(x= "", y = n, fill = Func.refGene)) + 
                geom_bar(stat = "identity", width = 1) + # , color = "white"
                coord_polar("y", start = 0) +
                # code for putting text labels on pie plot:
                # geom_text(aes(y = n, label = Func.refGene), color = "white") +
                # library("ggrepel")
                # geom_text_repel(aes(y = n, label = Func.refGene),
                #                 # color = "white",
                #                 show.legend = FALSE, 
                #                 # segment.alpha = 0, 
                #                 min.segment.length = 0,
                #                 force = 5) +
                # TODO: need to tweak this code so that it looks alright
                theme_bw() +
                theme(panel.grid.minor = element_blank(), 
                      panel.grid.major = element_blank(), 
                      panel.border = element_blank(), 
                      axis.ticks = element_blank(), 
                      axis.text = element_blank(), 
                      axis.title = element_blank()) +
                ggtitle("Target Types")
            
            plots[['types']] <- list(
                plot = variant_types_pie,
                subheader = 'Variant Types'
            )
            
            # ~~~~~~~~~~ VARIANT COUNT BARPLOT ~~~~~~~~~~~~~~~ # 
            counts_barplot <- ggplot(data = df, aes(x = Sample)) +
                geom_bar(stat="count") + 
                ggtitle("Number of Variants") + 
                coord_flip() +
                theme_bw() +
                theme(panel.grid.minor = element_blank()) 
            
            plots[['counts']] <- list(
                plot = counts_barplot,
                subheader = 'Counts'
            )
            # ~~~~~~~~~~~~~~~~~~~~~~~~~ # 
            qual_boxplot1 <- ggplot(data = df, aes(x=Sample, y=ABQ, fill=Sample)) + 
                geom_boxplot() + 
                guides(fill=FALSE) + 
                coord_flip() + 
                ggtitle("Variant Base Quality Scores") + 
                theme_bw() +
                theme(panel.grid.minor = element_blank())
            
            plots[['qual1']] <- list(
                plot = qual_boxplot1,
                subheader = 'Average quality of variant-supporting bases'
            )
            # ~~~~~~~~~~~~~~~~~~~~~~~~~ # 
            qual_boxplot2 <- ggplot(data = df, aes(x=Sample, y=RBQ, fill=Sample)) + 
                geom_boxplot() + 
                guides(fill=FALSE) + 
                coord_flip() + 
                ggtitle("Reference Base Quality Scores") + 
                theme_bw() +
                theme(panel.grid.minor = element_blank())
            
            plots[['qual2']] <- list(
                plot = qual_boxplot2,
                subheader = 'Average quality of reference-supporting bases'
            )
            # ~~~~~~~~~~~~~~~~~~~~~~~~~ # 
            depth_boxplot <- ggplot(data = df, aes(x=Sample, y=DP, fill=Sample)) + 
                geom_boxplot() + 
                guides(fill=FALSE) + 
                coord_flip() + 
                ggtitle("Depth of Coverage") + 
                theme_bw() +
                theme(panel.grid.minor = element_blank())
            
            plots[['dp']] <- list(
                plot = depth_boxplot,
                subheader = 'Depth of Coverage'
            )
            # ~~~~~~~~~~~~~~~~~~~~~~~~~ # 
            frequency_boxplot <- ggplot(data = df, aes(x=Sample, y=FREQ, fill=Sample)) + 
                geom_boxplot() + 
                guides(fill=FALSE) + 
                coord_flip() + 
                ggtitle("Allele Frequency") + 
                theme_bw() +
                theme(panel.grid.minor = element_blank())
            
            plots[['af']] <- list(
                plot = frequency_boxplot,
                subheader = 'Allele Frequency'
            )
            # ~~~~~~~~~~ VARIANT KARYOTYPE ~~~~~~~~~~~~~~~ # 
            kar <- karyotype + 
                geom_point(data = df, 
                           aes(x = as.numeric(chromosome),
                               y = Start), color = "red") +
                ggtitle("VarScan2 Variant Karotype") +
                theme(legend.position="none")
            
            plots[['kar']] <- list(
                plot = kar,
                subheader = 'Variant Karotype'
            )
        }
    }
    return(plots)
}

# ~~~~~~~ RUN ~~~~~~~ #
varscan2_indel_plots <- make_varscan2_indel_plots(annotations[which(annotations[["VariantCaller"]] == 'VarScan2' & annotations[["VariantCallerType"]] == "indel" ), ])
df <- annotations[which(annotations[["VariantCaller"]] == 'VarScan2' & annotations[["VariantCallerType"]] == "indel" ), ]
```

```{r, results='asis'}
# print any results found
if (length(varscan2_indel_plots) < 1){
    cat('\n\n- __No results found__\n\n')
} else {
    
    for(item in varscan2_indel_plots){
        subplot<- item[["plot"]]
        subheader <- item[["subheader"]]
        
        cat(sprintf('\n\n### %s\n\n', subheader))
        print(subplot)
    }
}
```


```{r}
# save environment to file
save.image(file="variants.Rdata",compress = TRUE)
# remove large objects that are no longer needed for report
rm(varscan2_snp_plots)
rm(varscan2_indel_plots)
rm(lofreq_plots)
rm(haplotypecaller_plots)
rm(mutect2_plots)
rm(lofreq_somatic_plots)
rm(annotations)
rm(subplot)
rm(item)
```
