```{r load_annotations}
# loads variant information from 'all_annotations.tsv' file
remove_vars(vars = c(
    ))

annotations_file <- input_files[grep(pattern = 'all_annotations.tsv', x = input_files)][1]
annotations <- read.delim(file = annotations_file, header = TRUE, sep = '\t', check.names = FALSE, na.strings = c('.'))
```

# Variants

## GATK HaplotypeCaller

Variant calling with the GATK `HaplotypeCaller` program [@GATKToolkit]

```{r haplotypecaller}
# ~~~~~~~ FUNCTIONS ~~~~~~~ #
make_haplotypecaller_plots <- function(df){
    plots <- list()
    if(is.data.frame(df)){
        if(nrow(df) > 0){
            # ~~~~~~~~~~~~~~~~~~~~~~~~~ # 
            haplotypecaller_counts <- ggplot(data = df, aes(x = Sample)) +
                geom_bar(stat="count") + 
                ggtitle("Number of Mutations") + 
                coord_flip() +
                theme_bw() +
                theme(panel.grid.minor = element_blank()) 
            
            plots[['counts']] <- list(
                plot = haplotypecaller_counts,
                subheader = 'Counts'
            )
            # ~~~~~~~~~~~~~~~~~~~~~~~~~ # 
            haplotypecaller_qual <- ggplot(data = df, aes(x=Sample, y=QUAL, fill=Sample)) + 
                geom_boxplot() + 
                guides(fill=FALSE) + 
                coord_flip() + 
                ggtitle("Quality Scores") + 
                theme_bw() +
                theme(panel.grid.minor = element_blank())
            
            plots[['qual']] <- list(
                plot = haplotypecaller_qual,
                subheader = 'Quality Scores'
            )
            # ~~~~~~~~~~~~~~~~~~~~~~~~~ # 
            haplotypecaller_dp <- ggplot(data = df, aes(x=Sample, y=DP, fill=Sample)) + 
                geom_boxplot() + 
                guides(fill=FALSE) + 
                coord_flip() + 
                ggtitle("Depth of Coverage") + 
                theme_bw() +
                theme(panel.grid.minor = element_blank())
            
            plots[['dp']] <- list(
                plot = haplotypecaller_dp,
                subheader = 'Quality Scores'
            )
            # ~~~~~~~~~~~~~~~~~~~~~~~~~ # 
            haplotypecaller_af <- ggplot(data = df, aes(x=Sample, y=AF, fill=Sample)) + 
                geom_boxplot() + 
                guides(fill=FALSE) + 
                coord_flip() + 
                ggtitle("Allele Frequency") + 
                theme_bw() +
                theme(panel.grid.minor = element_blank())
            
            plots[['af']] <- list(
                plot = haplotypecaller_af,
                subheader = 'Allele Frequency'
            )
        }
    }
    return(plots)
}

# ~~~~~~~ RUN ~~~~~~~ #
haplotypecaller_plots <- make_haplotypecaller_plots(annotations[annotations[["VariantCaller"]] == 'HaplotypeCaller', ])
```

```{r haplotypecaller_print, results='asis'}
# print any results found
if (length(haplotypecaller_plots) < 1){
    cat('\n\n- __No results found__\n\n')
} else {
    for(item in haplotypecaller_plots){
        subplot<- item[["plot"]]
        subheader <- item[["subheader"]]
        
        cat(sprintf('\n\n### %s\n\n', subheader))
        print(subplot)
    }
}
```

## LoFreq

Variant calling results from the LoFreq program [@LoFreq]

```{r lofreq}
# ~~~~~~~ FUNCTIONS ~~~~~~~ #
make_lofreq_plots <- function(df){
    plots <- list()
    if(is.data.frame(df)){
        if(nrow(df) > 0){
            # ~~~~~~~~~~~~~~~~~~~~~~~~~ # 
            lofreq_counts <- ggplot(data = df, aes(x = Sample)) +
                geom_bar(stat="count") + 
                ggtitle("Number of Mutations") + 
                coord_flip() +
                theme_bw() +
                theme(panel.grid.minor = element_blank()) 
            
            plots[['counts']] <- list(
                plot = lofreq_counts,
                subheader = 'Counts'
            )
            # ~~~~~~~~~~~~~~~~~~~~~~~~~ # 
            lofreq_qual <- ggplot(data = df, aes(x=Sample, y=QUAL, fill=Sample)) + 
                geom_boxplot() + 
                guides(fill=FALSE) + 
                coord_flip() + 
                ggtitle("Quality Scores") + 
                theme_bw() +
                theme(panel.grid.minor = element_blank())
            
            plots[['qual']] <- list(
                plot = lofreq_qual,
                subheader = 'Quality Scores'
            )
            # ~~~~~~~~~~~~~~~~~~~~~~~~~ # 
            lofreq_dp <- ggplot(data = df, aes(x=Sample, y=DP, fill=Sample)) + 
                geom_boxplot() + 
                guides(fill=FALSE) + 
                coord_flip() + 
                ggtitle("Depth of Coverage") + 
                theme_bw() +
                theme(panel.grid.minor = element_blank())
            
            plots[['dp']] <- list(
                plot = lofreq_dp,
                subheader = 'Quality Scores'
            )
            # ~~~~~~~~~~~~~~~~~~~~~~~~~ # 
            lofreq_af <- ggplot(data = df, aes(x=Sample, y=AF, fill=Sample)) + 
                geom_boxplot() + 
                guides(fill=FALSE) + 
                coord_flip() + 
                ggtitle("Allele Frequency") + 
                theme_bw() +
                theme(panel.grid.minor = element_blank())
            
            plots[['af']] <- list(
                plot = lofreq_af,
                subheader = 'Allele Frequency'
            )
        }
    }
    return(plots)
}

# ~~~~~~~ RUN ~~~~~~~ #
lofreq_plots <- make_lofreq_plots(annotations[annotations[["VariantCaller"]] == 'LoFreq', ])
```

```{r plot_variants, results='asis'}
# print any results found
if (length(lofreq_plots) < 1){
    cat('\n\n- __No results found__\n\n')
} else {
    
    for(item in lofreq_plots){
        subplot<- item[["plot"]]
        subheader <- item[["subheader"]]
        
        cat(sprintf('\n\n### %s\n\n', subheader))
        print(subplot)
    }
}
```

